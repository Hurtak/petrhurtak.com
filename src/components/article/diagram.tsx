import stripIndent from "strip-indent";

import { colors, gridCss } from "../../styles";

export const Diagram = ({ children }: { children: string }) => {
  return (
    <>
      <div>{stripIndent(children)}</div>

      <style jsx>{`
        div {
          background: ${colors.gray};
          border: 1px solid ${colors.grayDark};
          border-radius: 2px;
          padding: ${gridCss(1)};
          font-size: 12px;
          font-family: monospace;
          white-space: pre;
        }
      `}</style>
    </>
  );
};
