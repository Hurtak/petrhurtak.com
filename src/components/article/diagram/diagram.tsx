import * as React from "react";
import stripIndent from "strip-indent";

import { colors, gridCss } from "../../../styles";

export const Diagram = ({ children }: { children: string }) => {
  return (
    <>
      <div>{trimEmptyEdgeLines(stripIndent(children))}</div>

      <style jsx>{`
        div {
          background: ${colors.gray};
          border: 1px solid ${colors.grayDark};
          border-radius: 2px;
          line-height: 1;
          padding: ${gridCss(1)};
          font-size: 14px;
          font-family: monospace;
          white-space: pre;
        }
      `}</style>
    </>
  );
};

export const trimEmptyEdgeLines = (string: string): string => {
  const lines = string.split("\n");

  return lines
    .filter((line, index) => {
      if (line.trim() === "" && index === 0) return false;
      if (line.trim() === "" && index === lines.length - 1) return false;

      return true;
    })
    .join("\n");
};
