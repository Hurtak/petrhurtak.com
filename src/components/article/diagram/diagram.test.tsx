import * as React from "react";
import renderer from "react-test-renderer";
import { describe, expect, test } from "vitest";

import { Diagram, trimEmptyEdgeLines } from "./diagram";

describe("Diagram", () => {
  test("Diagram component snapshot", () => {
    const c = (
      <Diagram>{`
        ┌─┬─┐
        │ │ │
        ├─┼─┤
        │ │ │
        └─┴─┘
      `}</Diagram>
    );

    expect(renderer.create(c).toJSON()).toMatchSnapshot();
  });
});

describe("trimEmptyEdgeLines", () => {
  test("trims the edge lines", () => {
    expect(
      trimEmptyEdgeLines(`
        x
      `),
    ).toMatchInlineSnapshot('"        x"');
  });

  test("trims only one line from each edge", () => {
    expect(
      trimEmptyEdgeLines(`

        x

      `),
    ).toMatchInlineSnapshot(`
      "
              x
      "
    `);
  });

  test("does not trim when edge lines are not empty", () => {
    expect(trimEmptyEdgeLines(` x `)).toMatchInlineSnapshot('" x "');
    expect(
      trimEmptyEdgeLines(
        `x
x
x`,
      ),
    ).toMatchInlineSnapshot(`
      "x
      x
      x"
    `);
  });
});
