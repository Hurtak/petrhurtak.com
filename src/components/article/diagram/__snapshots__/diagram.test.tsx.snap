// Vitest Snapshot v1

exports[`Diagram > Diagram component snapshot 1`] = `
[
  <div>
    ┌─┬─┐
│ │ │
├─┼─┤
│ │ │
└─┴─┘
  </div>,
  <style
    jsx={true}
  >
    
        div {
          background: #f3f3f3;
          border: 1px solid #dadada;
          border-radius: 2px;
          line-height: 1;
          padding: 8px;
          font-size: 14px;
          font-family: monospace;
          white-space: pre;
        }
      
  </style>,
]
`;
